src/translate/pddl_parser/parsing_functions.py

```diff
@@ -1,31 +1,142 @@
+import contextlib
 import sys
 
 import graph
 import pddl
+from .parse_error import ParseError
 
+TYPED_LIST_SEPARATOR = "-"
 
-def parse_typed_list(alist, only_variables=False,
-                     constructor=pddl.TypedObject,
+SYNTAX_LITERAL = "(PREDICATE ARGUMENTS*)"
+SYNTAX_LITERAL_NEGATED = "(not (PREDICATE ARGUMENTS*))"
+SYNTAX_LITERAL_POSSIBLY_NEGATED = f"{SYNTAX_LITERAL} or {SYNTAX_LITERAL_NEGATED}"
+
+SYNTAX_PREDICATE = "(PREDICATE_NAME [VARIABLE [- TYPE]?]*)"
+SYNTAX_PREDICATES = f"(:predicates {SYNTAX_PREDICATE}*)"
+SYNTAX_FUNCTION = "(FUNCTION_NAME [VARIABLE [- TYPE]?]*)"
+SYNTAX_ACTION = "(:action NAME [:parameters PARAMETERS]? " \
+                "[:precondition PRECONDITION]? :effect EFFECT)"
+SYNTAX_AXIOM = "(:derived PREDICATE CONDITION)"
+SYNTAX_GOAL = "(:goal GOAL)"
+
+SYNTAX_CONDITION_AND = "(and CONDITION*)"
+SYNTAX_CONDITION_OR = "(or CONDITION*)"
+SYNTAX_CONDITION_IMPLY = "(imply CONDITION CONDITION)"
+SYNTAX_CONDITION_NOT = "(not CONDITION)"
+SYNTAX_CONDITION_FORALL_EXISTS = "({forall, exists} VARIABLES CONDITION)"
+
+SYNTAX_EFFECT_FORALL = "(forall VARIABLES EFFECT)"
+SYNTAX_EFFECT_WHEN = "(when CONDITION EFFECT)"
+SYNTAX_EFFECT_INCREASE = "(increase (total-cost) ASSIGNMENT)"
+
+SYNTAX_EXPRESSION = "POSITIVE_NUMBER or (FUNCTION VARIABLES*)"
+SYNTAX_ASSIGNMENT = "({=,increase} EXPRESSION EXPRESSION)"
+
+SYNTAX_DOMAIN_DOMAIN_NAME = "(domain NAME)"
+SYNTAX_TASK_PROBLEM_NAME = "(problem NAME)"
+SYNTAX_TASK_DOMAIN_NAME = "(:domain NAME)"
+SYNTAX_METRIC = "(:metric minimize (total-cost))"
+
+
+CONDITION_TAG_TO_SYNTAX = {
+    "and": SYNTAX_CONDITION_AND,
+    "or": SYNTAX_CONDITION_OR,
+    "imply": SYNTAX_CONDITION_IMPLY,
+    "not": SYNTAX_CONDITION_NOT,
+    "forall": SYNTAX_CONDITION_FORALL_EXISTS,
+}
+
+
+class Context:
+    def __init__(self):
+        self._traceback = []
+
+    def __str__(self) -> str:
+        return "\n\t->".join(self._traceback)
+
+    def error(self, message, item=None, syntax=None):
+        error_msg = f"{self}\n{message}"
+        if syntax:
+            error_msg += f"\nSyntax: {syntax}"
+        if item:
+            error_msg += f"\nGot: {item}"
+        raise ParseError(error_msg)
+
+    def expected_word_error(self, name, *args, **kwargs):
```

Comment:
I agree that it would be nicer to have a function do the check and yielding the error (if necessary). This applies to all three cases `expected_word_error`, `expected_list_error`, and `expected_name_block_error`. I also checked the calls to these functions and they all have similar if-clauses preceding them, so it shouldn't be a problem to unify this in a function.

If I understood Malte's comment in the daily sprint meeting correctly, they should preferrably be named `check...` rather than `assert...` because assert should be reserved for code errors rather than user input errors.